added parallax effect workshop as a part of workshop-bounty program (#1339)
hackclub/hackclub: workshops/parallax_effect/README.md

---
name: 'Parallax Effect'
description: 'Creating greater user experience on web using vanilla javascript'
author: '@giridhar7632'
---

> # Parallax Effect

Reviewer: Could we have a more exciting title?

Reviewer: ^^^ following up

Author: I have no idea about the title
I am thinking about it. But no use
It's just an effect. I can't find any interesting title.😔

Author: Please let me know if you have anything in your mind

Reviewer 2: I kinda like “parallax effect”, it sounds cool

Author: Okay

> **using vanilla javaScript**

Reviewer: ?

> One of the most popular effects of the modern web experience is the **parallax effect**. We will be building basic-starter parallax effect 

Reviewer: Check the grammar / phrasing on this sentence

Author: I did my best. 
Please let me know, if there are any changes to be done

Author: Is the grammar okay?

> You should have some beginner-intermediate knowledge about HTML, CSS and some javaScript.

Reviewer: ```suggestion
To understand this workshop you will need to be familiar with HTML, CSS and Javascript.
```

We are going to make something like this

<a href="https://total-parallax.giridharhackclu.repl.co/">![Final REsult](https://cloud-1z03ndewk.vercel.app/ezgif.com-gif-maker__2_.gif)</a>

Here's the [live demo](https://total-parallax.giridharhackclu.repl.co/), and [source code](https://repl.it/@Giridharhackclu/total-parallax#index.html)

## Parallax

Parallax movement is when things move at different speeds relative to each other. This effect can be used to create an illusion of depth. 
We will actually experience this effect while scrolling a webpage. It's called Parallax scrolling.

The technique is popular in many places including video games, where it’s usually seen in the layered backgrounds.

> But what exactly makes the illusion?<br>

Reviewer: ```suggestion
But what exactly makes the illusion?

```

![Final Result](https://cloud-1z03ndewk.vercel.app/ezgif.com-gif-maker__2_.gif)

That's it! You added both `vertical-parallax` and `horizontal-parallax` effects. Well done.

## Hacking

Your creativity is not restricted to only this example. You can create different parallax effects. 
1) In this example, I focused only on scrolling vertically. You can use <a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollX">`scrollX`</a> for creating parallax while scrolling horizontally.
1) You can use <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/translate">`transform: translate(X,Y);`</a> or <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/translate3d">`transform: translate3d(X,Y,Z);`</a> for creating different effects.

You can add parallax-effect to your websites for creating better user experience and display your content more interactively.

Make use of creativity and create different parallax effects. Share it with me on slack [@giridhar](https://hackclub.slack.com/team/U013E6KE9UJ), I'd love to here from you!

## Inspiration

These are some projects made using parallax-effect. 

> * Example-1 [demo](https://exampleone.giridharhackclu.repl.co/). [source code](https://repl.it/@Giridharhackclu/exampleone#index.html)

Reviewer: Add a brief explanation about this

We can see how the parallax scrolling works in this example [here](https://codepen.io/samdbeckham/full/OPXPNp).

<a href="https://codepen.io/samdbeckham/full/OPXPNp" target="_blank"><img src="https://cloud-kt149rdr8.vercel.app/ezgif.com-add-text.gif" alt="parallax scrolling"/></a>

^As you scroll, each of the layers moves at different speeds.

## Advantages of Parallax Effect

If you want to make your website amazing, it's not enough to show static elements. You have to amaze your visitors and make them want to explore your site, otherwise they can leave fast. You have to show them some cool things and some short visual surprises. The Parallax effect makes it possible. Moving components on scrolling or on hovering will make them interested.
You can create something interesting using parallax effect of your choice.

## Getting started

Practically a parallax is composed of more than one layer in parallel, moving along on scroll at different speeds, giving us the feeling that they're at different distances.

### Setup

We will use [Repl.it](https://repl.it) to build this project.

> > Repl.it is an online code editor. It's similar to Google Docs, but has some important features that make it much better for typing code than a regular text editor.

Reviewer 2: I would remove the `>` here, it looks a little weird.

Author: okay!

Fork this repl and explore! [here](https://repl.it/@Giridharhackclu/parallax-starter#index.html).

> It contains three `div` elements with classes `layer l1`, `layer l2` and `layer l3` added with some basic styles. Go ahead! use your creativity and customise the styles.

Reviewer 2: > Go ahead! use your creativity and customise the styles.

This is a great prompt! But at this stage of the workshop, the person doing the workshop barely knows what’s going on since we’re just at the beginning, so I think this prompt is best saved for the end.

Author: okay!

> <a href="https://repl.it/@Giridharhackclu/parallax-starter#index.html"><img src="https://cloud-6z92hbh0u.vercel.app/screenshot_2020-09-22_171732.png" style="{border-radius: 3px;}" alt="#index.html"></a>

Reviewer 2: Instead of an image, could you make this a code block?

Author: I kept an image, so that they can see what actually is in the repl

Author: I am changing it into code block

For applying parallax effect, the thing we need to do is select the layer and change it's speed while scrolling. That's it!

Let's start doing it!

Go to [script.js](https://repl.it/@Giridharhackclu/parallax-starter#script.js) and add the following <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions">`function`</a>.

```javascript
function parallax(layer, distance, speed) {
  const item = document.querySelector(layer)
  item.style.transform = 'translateY(' + -distance * speed + 'px)'
}
```

> This `function` has three arguments `layer`- the layer you want to add parallax, `distance`- how much we scroll and `speed`- the required speed change. We will add parallax for multiple layers. That is the reason we made a function to recall it for different layers.

Reviewer 2: I think this paragraph is a little confusing as a paragraph. Maybe make it into a list.

This `function` has three arguments:
* `layer`: the layer you want to add the parallax effect to
* `distance`: how much we scroll
* `speed`: the required speed change

We will add parallax...

Author: I really like this suggestion. Thank you

> Then we will get the element(layer) required using `querySelector`. You can use any JavaScript DOM selector as your wish. Then translate the `Y-offset` of the element using CSS `transform: translateY();` to change the speed of that layer.

Reviewer 2: Not sure what this paragraph is trying to say. To me it looks like they’re instructions and I’m supposed to do something, but I’m not sure what I’m supposed to do. Clarify here maybe

Reviewer 2: Oh I see now that it’s more of a description of the previous code block. In that case, I would keep this as part of the previous paragraph

Author: Okay:smile:

The positive value into `translateY()` translates downwards and negative value translates upwards. Discover more about `translateY()` [here](https://developer.mozilla.org/en-US/docs/web/css/transform-function/translateY).

<a href="https://cloud-7jtog4dbv.vercel.app/screenshot_2020-09-22_184214.png"><img src="https://cloud-7jtog4dbv.vercel.app/screenshot_2020-09-22_184214.png" alt="translateY() example" height="250px" width="auto"/></a>

## addEventListener

Now call this `parallax` function while scrolling. Add an [event listener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) <a href="https://developer.mozilla.org/en-US/docs/Web/API/Document/scroll_event">`scroll`</a> . The function given as argument to event-listener gets executed.

```javascript
document.addEventListener('scroll', () => {})
```

> > **scrollY:** The read-only scrollY property of the Window interface returns the number of pixels that the document is currently **scrolled vertically**. This value is subpixel precise in modern browsers, meaning that it isn't necessarily a whole number. You can get the number of pixels the document is **scrolled horizontally** from the **scrollX** property. - [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollY).

Reviewer 2: I love this! Looks like optional docs. I think this “>” formatting really works here.

Author: yeah

So finally check the output and scroll.

Your final page looks something [like this](https://parallax-effect.giridharhackclu.repl.co/).

<a href="https://parallax-effect.giridharhackclu.repl.co/"><img src="https://cloud-qld8y6jii.vercel.app/final_output.gif" alt="final result"/></a>

Try changing the speed to `1`, observe what happens to each layer and think why it happens. You can check the answer [here](#Speed-1).

That's it! You can add any number of layers for the document and call the `parallax` function. This is the _parallax effect_.
These are the main basic principle of Parallax effect. You can modify the function `parallax` and get the result you want.

Let's try another method of applying parallax.

## Horizontal Parallax

Now add another two `div` tags with classes `layer l4` and `layer l5` to `index.html`

```html
<div class="layer l4">Right</div>
<div class="layer l5">Left</div>
```

In horizontal parallax, the elements in the page move horizontally as you scroll vertically.

> Add the following `function` , which causes horizontal motion.

Reviewer 2: specify “to your `script.js` file”

Make use of your creativity and make different parallax effects. Share it with me on slack [@giridhar](https://hackclub.slack.com/team/U013E6KE9UJ), I'd love to here from you!

## Inspiration

These are some projects made using parallax-effect.

- Example-1 [demo](https://exampleone.giridharhackclu.repl.co/). [source code](https://repl.it/@Giridharhackclu/exampleone#index.html).
  Real application of parallax effect.
- Example-2 [demo](https://exampletwo.giridharhackclu.repl.co/). [source code](https://repl.it/@Giridharhackclu/exampletwo#index.html).
  _Zoom on scroll_ - In this example, width of elements changes as you scroll.
- Example-3 [demo](https://horizontal-parallax-effect.giridharhackclu.repl.co/). [source code](https://repl.it/@Giridharhackclu/horizontal-parallax-effect#index.html).
  _Mouse-move parallax_ - As you move mouse-pointer, the `innerHeight` and the `innerWidth` of the elements changes.

Check this article for awesome next-level parallax examples - [article](https://www.awwwards.com/30-great-websites-with-parallax-scrolling.html)

Hope you love this workshop! :v:

> ## Speed-1

Reviewer 2: Any reason why this is at the very end? It feels like a part of the workshop got bitten off and spit out at the end. Should this be higher up?

Author: I got inspired this form sam's workshop
He asked a question and linked the answer at the last
so I did in the same way
No problem, I will change it!

---
name: 'Parallax Effect'
description: 'Creating greater user experience on web using javaScript'
author: '@giridhar7632'
---

# Parallax Effect

> **creating illusion of depth using javaScript**

Reviewer: this is a weird format

Author: I thought It would be nice if there is a sub-title
I have deleted it.

One of the most popular effects of modern web experience is **the parallax effect**. We are going to build a basic project to understand how the parallax effect works.
To understand this workshop you will need to be familiar with HTML, CSS and JavaScript.

We are going to make something like this

> <a href="https://total-parallax.giridharhackclu.repl.co/">![Final REsult](https://cloud-1z03ndewk.vercel.app/ezgif.com-gif-maker__2_.gif)</a>

Reviewer: ```suggestion
<a href="https://total-parallax.giridharhackclu.repl.co/">![Final Result](https://cloud-1z03ndewk.vercel.app/ezgif.com-gif-maker__2_.gif)</a>
```
